pkg/dtest/k3s.go

```diff
@@ -0,0 +1,312 @@
+	"replicasets.extensions",
+	"helmcharts.helm.cattle.io",
+	"addons.k3s.cattle.io",
+	"listenerconfigs.k3s.cattle.io",
+	"ingresses.networking.k8s.io",
+	"networkpolicies.networking.k8s.io",
+	"runtimeclasses.node.k8s.io",
+	"poddisruptionbudgets.policy",
+	"podsecuritypolicies.policy",
+	"clusterrolebindings.rbac.authorization.k8s.io",
+	"clusterroles.rbac.authorization.k8s.io",
+	"rolebindings.rbac.authorization.k8s.io",
+	"roles.rbac.authorization.k8s.io",
+	"priorityclasses.scheduling.k8s.io",
+	"csidrivers.storage.k8s.io",
+	"csinodes.storage.k8s.io",
+	"storageclasses.storage.k8s.io",
+	"volumeattachments.storage.k8s.io",
+}
+
+func isK3sReady() bool {
+	kubeconfig := getKubeconfigPath()
+
+	if kubeconfig == "" {
+		return false
+	}
+
+	cmd := supervisor.Command(prefix, "kubectl", "--kubeconfig", kubeconfig, "api-resources", "-o", "name")
+	output, err := cmd.Capture(nil)
+	if err != nil {
+		return false
+	}
+	resources := make(map[string]bool)
+	for _, line := range strings.Split(output, "\n") {
+		resources[strings.TrimSpace(line)] = true
+	}
+
+	for _, req := range requiredResources {
+		_, exists := resources[req]
+		if !exists {
+			return false
+		}
+	}
+
+	return true
+}
+
+const k3sConfigPath = "/etc/rancher/k3s/k3s.yaml"
+
+// GetKubeconfig returns the kubeconfig contents for the running k3s
+// cluster as a string. It will return the empty string if no cluster
+// is running.
+func GetKubeconfig() string {
+	if !isKubeconfigReady() {
+		return ""
+	}
+
+	id := tag2id("k3s")
+
+	if id == "" {
+		return ""
+	}
+
+	cmd := supervisor.Command(prefix, "sh", "-c", fmt.Sprintf("docker cp \"%s:%s\" - | tar -xO", id, k3sConfigPath))
+	kubeconfig := cmd.MustCapture(nil)
+	kubeconfig = strings.ReplaceAll(kubeconfig, "localhost:6443", fmt.Sprintf("%s:%s", dockerIP(), k3sPort))
+	return kubeconfig
+}
+
+func getKubeconfigPath() string {
+	id := tag2id("k3s")
+
+	if id == "" {
+		return ""
+	}
+
+	user, err := user.Current()
+	if err != nil {
+		panic(err)
+	}
+
+	kubeconfig := fmt.Sprintf("/tmp/dtest-kubeconfig-%s-%s.yaml", user.Username, id)
+	contents := GetKubeconfig()
+
+	err = ioutil.WriteFile(kubeconfig, []byte(contents), 0644)
+
+	if err != nil {
+		panic(err)
+	}
+
+	return kubeconfig
+}
+
+const dtestRegistry = "DTEST_REGISTRY"
+const registryPort = "5000"
+
+// RegistryUp will launch if necessary and return the docker id of a
+// container running a docker registry.
+func RegistryUp() string {
+	return dockerUp("registry",
+		"-p", fmt.Sprintf("%s:6443", k3sPort),
+		"-p", fmt.Sprintf("%s:%s", registryPort, registryPort),
+		"-e", fmt.Sprintf("REGISTRY_HTTP_ADDR=0.0.0.0:%s", registryPort),
+		"registry:2")
+}
+
+func dockerIP() string {
+	return "localhost"
+}
+
+// DockerRegistry returns a docker registry suitable for use in tests.
+func DockerRegistry() string {
+	registry := os.Getenv(dtestRegistry)
+	if registry != "" {
+		return registry
+	}
+
+	RegistryUp()
+
+	return fmt.Sprintf("%s:%s", dockerIP(), registryPort)
```

Comment:
Use `net.JoinHostPort(dockerIP(), registryPort)` instead of `fmt.Sprintf("%s:%s", dockerIP(), registryPort)`.  If `dockerIP()` is ever modified to return an IPv6 address, the Sprintf solution won't behave correctly.